### ISSUE TITLE
A sufficiently large number seems to disable learning

### ISSUE THREAD

**Issue by amosr** [JohnLangford/vowpal_wabbit#651]
From what I can tell, if the first row has a number larger than 18446745900000000000, learning just doesn't happen.

Given the following input file:
```
1 | out:1       ignore:18446745900000000000
1 | out:1
1 | out:1
1 | out:1
1 | out:1
1 | out:1
1 | out:1
1 | out:1
```
Now, I'd expect to ignore "ignore" and just use out, or maybe a constant 1.

However when you run it, it just seems to give up and keep the model at 0.
```
$ vw data4/test
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = data4/test
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
1.000000   1.000000          1      1.0     1.0000   0.0000        2
1.000000   1.000000          2      2.0     1.0000   0.0000        2
1.000000   1.000000          4      4.0     1.0000   0.0000        2
1.000000   1.000000          8      8.0     1.0000   0.0000        2

finished run
number of examples per pass = 8
passes used = 1
weighted example sum = 8
weighted label sum = 8
average loss = 1
best constant = 1
best constant's loss = 0
total feature number = 16
```

And if you output the model file, we just get
```
Version 7.10.0
Min label:0.000000
Max label:1.000000
bits:18
0 pairs:
0 triples:
lda:0
0 ngram:
0 skip:
options:
:0
```

I should note that I haven't tried with latest master yet, only the version I got off homebrew (7.10.0). I figure this is worth documenting either way though, as my searching didn't turn up anything about it.

**Comment by JohnLangford** [comment]
When I run this example, I see nan cropping up (as below).

Tracing, the problem is that large*large = inf and the normalized update 
makes this inf*0 = nan.

You should be warned that something is going wrong.  Why don't you see 
NAN in the printout?

-John

10:36AM humpty-219: ./vw temp
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = temp
num sources = 1
average  since         example        example  current  current current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0   1.0000 0.0000        3
NAN prediction in example 2, forcing 0.000000
1.000000 1.000000            2            2.0   1.0000 0.0000        2
NAN prediction in example 3, forcing 0.000000
NAN prediction in example 4, forcing 0.000000
1.000000 1.000000            4            4.0   1.0000 0.0000        2
NAN prediction in example 5, forcing 0.000000
NAN prediction in example 6, forcing 0.000000
NAN prediction in example 7, forcing 0.000000
NAN prediction in example 8, forcing 0.000000
1.000000 1.000000            8            8.0   1.0000 0.0000        2

finished run
number of examples per pass = 8
passes used = 1
weighted example sum = 8.000000
weighted label sum = 8.000000
average loss = 1.000000
best constant = 1.000000
best constant's loss = 0.000000
total feature number = 17

**Comment by amosr** [comment]
Wow, thanks for the quick reply.

I'm not sure why, but I'm definitely not getting NAN errors. Could it be the version I'm running? I'll try compiling master later today and see if that gives me the errors.

**Comment by amosr** [comment]
Hi,
I still don't know why I'm not getting the NAN errors in finalize_predict. However, in my case the actual prediction was 0 but the pred_per_update was NAN.
So I added a similar warning to pred_per_update in pull request #667, if you are interested.

Thanks

**Comment by JohnLangford** [comment]
Closing this because we have an outstanding pull request on the same issue.
